File: 141. Linked List Cycle/linked-list-cycle.c
// 141. Linked List Cycle (7/12/54152)
// Runtime: 14 ms (46.46%) Memory: 8.11 MB (0.00%) 

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle (struct ListNode* head) {
    if (head == NULL || head->next == NULL) {
        return false;
    }
    
    struct ListNode* tail = head;
    
    while (head->next && head->next->next) {
        head = head->next->next;
        tail = tail->next;

        if (head == tail) {
            return true;
        }
    }
    
    return false;
}

File: 141. Linked List Cycle/linked-list-cycle.js
// 141. Linked List Cycle (9/6/53727)
// Runtime: 119 ms (17.28%) Memory: 42.31 MB (93.69%) 

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
function hasCycle(head) {
    const set = new Set();
    let tail = head;
    while(tail) {
        set.add(tail);
        if(set.has(tail.next)) return true;
        tail = tail.next;
    }
    return false;
};